app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-core.js

```diff
@@ -218,6 +218,28 @@ window.GOVUK.analyticsGa4 = window.GOVUK.analyticsGa4 || {};
             domainsArrays.push(domainWithoutWww)
           }
         }
+      },
+
+      // The index and index_total properties are usually set by the application's templating engine however in instances where this isn't possible,
+      // we use JavaScript to calculate the index_total and append it to data-ga4-link and add an index property to each link
+      setIndexes: function (module) {
+        var linksToBeIndexed = []
+
+        // Only index links that are not search results
+        // querySelectorAll(a:not['data-ga4-ecommerce-path']) would have been used if not for lack of support in older browsers
+        for (var i = 0; i < module.querySelectorAll('a').length; i++) {
+          if (!module.querySelectorAll('a')[i].getAttribute('data-ga4-ecommerce-path')) {
+            linksToBeIndexed.push(module.querySelectorAll('a')[i])
+          }
+        }
+
+        var ga4LinkData = JSON.parse(module.getAttribute('data-ga4-link'))
+        ga4LinkData.index_total = linksToBeIndexed.length
+        module.setAttribute('data-ga4-link', JSON.stringify(ga4LinkData))
+
+        for (var j = 0; j < linksToBeIndexed.length; j++) {
+          linksToBeIndexed[j].setAttribute('data-ga4-index', j + 1)
+        }
```

Comment:
I think this second loop might not be necessary. I've rewritten this function a little, how does this look?

```Javascript
setIndexes: function (module) {
  var links = module.querySelectorAll('a')
  var total_links = 0
  for (var i = 0; i < links.length; i++) {
    var link = links[i]
    // Only index links that are not search results
    if (!link.getAttribute('data-ga4-ecommerce-path')) {
      total_links++
      link.setAttribute('data-ga4-index', i + 1)
    }
  }

  var ga4LinkData = JSON.parse(module.getAttribute('data-ga4-link'))
  ga4LinkData.index_total = total_links
  module.setAttribute('data-ga4-link', JSON.stringify(ga4LinkData))
}
```

Reply:
Thanks Andy, I've just tested this and I think I might need both loops - the problem is that if I only have the one loop, the `index` property will be set incorrectly. So in the code above, it sets the `data-ga4-index` attribute correctly but the index would’ve kept incrementing as it’s taking all links into account. For example, if there are 10 links but the first 7 are search results, clicking on the first non-search result link would look like:

```
index: 8
index_total: 3 
```

Reply:
Ah of course. Would it fix it to amend `link.setAttribute('data-ga4-index', i + 1)` to `link.setAttribute('data-ga4-index', total_links + 1)`?

Reply:
Very good point, thanks Andy - although I think we might be able to just set it to total_links

Reply:
Ah of course it's already being incremented on the previous line. Excellent.